### PR TITLE
Reduce to a single lock for RefreshSummaryAction and Summarization

### PR DESCRIPTION
**Pull request by NicholasCouri**
Due to the asynchronicity of our events, the summarization can start just before the refresh of the latest summary gets invoked.
The call refreshSummaryAckLock (runningSummarizer.ts) happens before the actual lock is taken by the call to lockedRefreshSummaryAckAction.  This change reduces to a single lock to address this problem. 
 
Additionally, we want the first summarizer to refresh the last summary, in case it has already been processed (summary collection), before the summarization begins.  The proposed solution makes sure we will refresh the latest summary before going into a loop of waiting for a new  SummaryAcks and also before the summarization starts.

### CODE REVIEW

**Comment by NicholasCouri** [comment]
> Can we simply bail out (exit) on # 3?

Can't we have a scenario where anther summarizer just executed a lastSummary and, we get the Summary Ack, after our summarizer has started ? Or just before it starts? We caught up and started the summarizer but got this new Summary Ack? 
If so, can should not bail out on #3 ?  Should we ?

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 1]
```diff
@@ -390,68 +390,29 @@ export class Summarizer extends EventEmitter implements ISummarizer {
...
 	};
 
-	private async handleSummaryAcks() {
-		let refSequenceNumber = this.runtime.deltaManager.initialSequenceNumber;
-		let ack: IAckedSummary | undefined;
+	/**
+	 * Responsible for receiving and processing all the summaryAcks.
+	 * @param lastAckRefNumber - Before initializing the summarization, if there is a last Ack to be processed, use it.
+	 */
+	private async handleSummaryAcks(lastAckRefNumber: number) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizer.ts:397]: Can we move this function into the RunningSummarizer so that all the handleSummaryAck logic is in one place. It's very hard to follow whats going on since its distributed in 2 different places.
All that is needed it to pass in `this.runtime.deltaManager.initialSequenceNumber` to running summarizer.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 2]
```diff
@@ -390,68 +390,29 @@ export class Summarizer extends EventEmitter implements ISummarizer {
...
+		let refSequenceNumber =
+			lastAckRefNumber > 0
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizer.ts:399]: If I understand this, this is saying that if there was an ack processed (by the running summarize) before this function is called, start waiting for ack that is newer that the one that was processed?
Could you please add comments here. It's difficult to follow other wise.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 3]
```diff
@@ -390,68 +390,29 @@ export class Summarizer extends EventEmitter implements ISummarizer {
...
+
+			summaryLogger.sendTelemetryEvent({
+				eventName: "handleSummaryAcks",
+				referenceSequenceNumber: refSequenceNumber,
+				lastAckRefNumber,
+			});
+
+			// Initialize ack with undefined if exception happens inside of waitSummaryAck on second iteration,
+			// we record undefined, not previous handles.
+			await this.summaryCollection.waitSummaryAck(refSequenceNumber);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizer.ts:414]: I think this await needs to move into the runningSummarizer.handleSummaryAck() function. That function assumes that an ack has already been awaited for and if that has not happened, it will end up returning -1 and this logic will break.

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 4]
```diff
@@ -216,6 +229,78 @@ export class RunningSummarizer implements IDisposable {
 		});
 	}
 
+	public async handleSummaryAck(): Promise<number> {
+		const lastAck: IAckedSummary | undefined = this.summaryCollection.latestAck;
+		let refSequenceNumber = -1;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:234]: As per my [other comment](https://github.com/microsoft/FluidFramework/pull/14142/files#r1106412267), this should not return -1. It should wait for ack and use its reference sequence number.

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 5]
```diff
@@ -429,16 +498,24 @@ export class RunningSummarizer implements IDisposable {
...
+				const summarizeResult = this.generator.summarize(
+					summarizeProps,
+					options,
+					cancellationToken,
+					resultsBuilder,
+				);
+				// ensure we wait till the end of the process
+				return summarizeResult.receivedSummaryAckOrNack;
+			},
+			() => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:512]: Just curious, why not do the before and after directly at the call site instead of inside the lockedSummaryAction function?

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 6]
```diff
@@ -353,41 +438,35 @@ export class RunningSummarizer implements IDisposable {
...
 	 * Assumes that caller checked upfront for lack of concurrent action (this.summarizingLock)
 	 * before calling this API. I.e. caller is responsible for either erroring out or waiting on this promise.
 	 * @param action - action to perform.
+	 * @param before - set of instructions to run before running the action.
+	 * @param after - set of instructions to run after running the action.
 	 * @returns - result of action.
 	 */
-	private async lockedSummaryAction<T>(action: () => Promise<T>) {
+	private async lockedSummaryAction<T>(
+		action: () => Promise<T>,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:466]: nit: Re-order the params to: before, action and after so it is more inline with how these will be processed.
Another option is to move action to the end (before, after, action) because the action can be a large block (as in trySumamrize) making it a little difficult to read what before and after are doing.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 1 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/summarizer.ts:397]: not even that as we pass the runtime.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 4 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: THere might not be an ack. Isn't it ? We can't wait forever here. if it is available, great, if not proceed.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 5 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:550]: no particular reason. the idea was to keep existing behavior of the lockdSummaryAction.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: I see. The very first time it runs, there may not be an ack so it should not wait.

My concern is that its assuming that this.summaryCollection.latestAck will be there when its called (after the first time it runs). The caller awaits the ack so this will be true today but it has the potential to go wrong if that code is modified.

I would rather pass in an ack to this function and it simply handles that ack. Then the two calll sites are:
1. During start - this.summaryCollection.lastAck is passed.
2. From handleSummaryAcks() - It will pass on the result of await this.summaryCollection.waitSummaryAck(refSequenceNumber).

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 7]
```diff
@@ -224,6 +240,105 @@ export class RunningSummarizer implements IDisposable {
...
+			refSequenceNumber++;
+		}
+		return refSequenceNumber;
+	}
+
+	/**
+	 * Responsible for receiving and processing all the summaryAcks.
+	 * @param lastAckRefNumber - Before initializing the summarization, if there is a last Ack to be processed, use it.
+	 */
+	private async handleSummaryAcks(lastAckRefNumber: number) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:320]: We should probably name this that reflects it will continuously wait for acks. Also, it sounds very similar to handleSummaryAck :-).
Maybe something like `processIncomingSummaryAcks`?

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 8]
```diff
@@ -76,8 +81,17 @@ export class RunningSummarizer implements IDisposable {
 			runtime,
 		);
 
+		// Before doing any heuristics or proceeding with its refreshing, lets refresh based on latest Summary ack if already available.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:84]: Can you please add a little more detail why this is done. Basically, if there are any acks that were received while this summarizer catches up, we want to refresh state from it before attempting a summary?

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 9]
```diff
@@ -224,6 +240,105 @@ export class RunningSummarizer implements IDisposable {
...
+				);
+			}
+			refSequenceNumber++;
+		}
+		return refSequenceNumber;
+	}
+
+	/**
+	 * Responsible for receiving and processing all the summaryAcks.
+	 * @param lastAckRefNumber - Before initializing the summarization, if there is a last Ack to be processed, use it.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:318]: The description of lastAckRefNumber is not clear. From my understanding, if this is >0, handleSummaryAcks() needs to wait for an ack for a summary that is newer than this ref seq number. Basically, we may already have processed an ack before this is called and so it should only process acks which are newer than that one.

Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:318]: nit: lastAckRefNumber -> lastAckRefSeq

**Comment by vladsud** [comment]
>> Can we simply bail out (exit) on # 3?

> Can't we have a scenario where anther summarizer just executed a lastSummary and, we get the Summary Ack, after our summarizer has started ? Or just before it starts? We caught up and started the summarizer but got this new Summary Ack?
If so, can should not bail out on # 3 ? Should we ?

We can, but last summarizer should prevent another summarizer from starting, even if its parent is gone. I.e., the logic that calculates who should be summarizer accounts for both main and summarizer client pair - the priority goes to selecting existing summarizer client, then to oldest main client.

While races will happen (due to no global synchronization around disconnects), they are rare (percent wise) and should be handled uniformly the same way.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 4 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: Not sure I follow -  it does not expect the latestAck to be there at all.  The idea is that only handleSummaryAcks will wait for the summary ack,

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: It does not expect it directly. But there are 2 issues that can happen:
1. If await latestAck is not called before calling this, it will unncessarily call refresh on container runtime which would end up in a no-op. The (unsaid) expected usage is that the caller makes sure latest ack is awaited. I am suggesting making that expectation explicit.
2. If there is no ack, it will return -1. The caller has to handle this scenario. handleSummaryAcks() does not handle this scenario because it knows there is an ack before it calls this function. But its not explicitly spelled out and so there are chances to the caller making mistakes.

I believe we are saying the same thing :-). I am just suggesting making things more explicit so that failures don't go unnoticed causing other issues.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 7 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:320]: Better indeed.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 9 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:318]: Tried to clarify in the next iteration.

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: At the very least, handleSummaryAcks() should assert that the return value from this function is not -1.

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 4 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:245]: I see. Added assert,

**Comment by NicholasCouri** [comment]
> If I'm correct (and maybe we need to add telemetry to validate that claim), then I think we should get rid of that loop. I'm Ok if we take time and move safely here, but let's either do that, or prove (via data / description of particular scenarios) where I'm wrong.

I've added telemetry to collect any Ack that we get after the initial one - just to find out how often it happens and then we proceed with the removal.
